### PR TITLE
Release v4.0.0 with pylint and flake8 support

### PR DESCRIPTION
**Pull request by pe-st**
pylint and flake8 are executed automatically in the Github build pipeline,
and the code was adapted to have a clean build,
e.g. using f-strings everywhere.

There is also some documentation about the supported Python versions

I'd like to gather your feedback and then release version 4.0.0 around the middle of august

I'm especially pinging @bxsx / @joetimmerman / @cristian5th / @rsjrny because of their open pull requests, but I'm happy for feedback from anyone.

### CODE REVIEW

**Review by bxsx — APPROVED** [review]
A ton of good work @pe-st ! Nice 👍

Have you considered adding `black` for auto-formatting?
I've also left a few comments—nothing serious though.

File: .pylintrc [hunk 1]
```diff
@@ -1,135 +1,163 @@
...
 load-plugins=
 
 # Pickle collected data for later comparisons.
 persistent=yes
 
-# Specify a configuration file.
-#rcfile=
+# Minimum Python version to use for version dependent checks. Will default to
+# the version used to run pylint.
+py-version=3.8
```
[the remark this answers is not in the document]
Comment (pe-st) [reply .pylintrc:83]: I think I should not set it at all so it determines the version dynamically from itself

**Comment by pe-st** [comment]
> Have you considered adding `black` for auto-formatting? I've also left a few comments—nothing serious though.

I'm experimenting with it. At least I will adopt some formatting from black...

File: filtering.py [hunk 2]
```diff
@@ -15,54 +17,52 @@ def read_exclude(file):
...
-        # except JSONDecodeError: only in Python3, use the more generic type instead
-        except ValueError:
+        except json.JSONDecodeError:
             obj = ""
 
     # Sanitize wrong formats
-    if not type(obj) is dict:
-        obj = dict()
+    if not isinstance(obj, dict):
+        obj = {}
```
[the remark this answers is not in the document]
Comment (pe-st) [reply filtering.py]: thx, applied!

File: gcexport.py [hunk 3]
```diff
@@ -1132,6 +1141,108 @@ def copy_details_to_summary(summary, details):
...
+    # Retrieve also the detail data from the activity (the one displayed on
+    # the https://connect.garmin.com/modern/activity/xxx page), because some
+    # data are missing from 'actvty' (or are even different, e.g. for my activities
+    # 86497297 or 86516281)
+    activity_details, details = fetch_details(actvty['activityId'], http_req_as_string)
+
+    extract = {}
+    extract['start_time_with_offset'] = offset_date_time(actvty['startTimeLocal'], actvty['startTimeGMT'])
+    elapsed_duration = details['summaryDTO']['elapsedDuration'] if 'summaryDTO' in details and 'elapsedDuration' in \
+                                                                   details['summaryDTO'] else None
```
[the remark this answers is not in the document]
Comment (pe-st) [reply gcexport.py]: thx, applied!

File: gcexport.py [hunk 4]
```diff
@@ -1132,6 +1141,108 @@ def copy_details_to_summary(summary, details):
...
+    extract['elapsed_duration'] = elapsed_duration if elapsed_duration else actvty['duration']
+    extract['elapsed_seconds'] = int(round(extract['elapsed_duration']))
+    extract['end_time_with_offset'] = extract['start_time_with_offset'] + timedelta(seconds=extract['elapsed_seconds'])
+
+    print('\t', extract['start_time_with_offset'].isoformat(), ', ', sep='', end='')
+    print(hhmmss_from_seconds(extract['elapsed_seconds']), ', ', sep='', end='')
+    if 'distance' in actvty and isinstance(actvty['distance'], (float)):
```
[the remark this answers is not in the document]
Comment (pe-st) [reply gcexport.py]: thx, applied!

File: gcexport.py [hunk 5]
```diff
@@ -1132,6 +1141,108 @@ def copy_details_to_summary(summary, details):
...
+            activity_measurements = http_req_as_string(f"{URL_GC_ACTIVITY}{actvty['activityId']}/details")
+            write_to_file(os.path.join(args.directory, f"activity_{actvty['activityId']}_samples.json"),
+                          activity_measurements, 'w',
+                          start_time_seconds)
+            samples = json.loads(activity_measurements)
+            extract['samples'] = samples
+        except HTTPError:
+            pass  # don't abort just for missing samples...
+            # logging.info("Unable to get samples for %d", actvty['activityId'])
+            # logging.exception(e)
```
[the remark this answers is not in the document]
Comment (pe-st) [reply gcexport.py]: Done. I don't remember why I commented them out, maybe the behaviour of the Garmin site was worse in the past?

File: README.md [hunk 6]
```diff
@@ -58,7 +58,7 @@ usage: gcexport.py [-h] [--version] [-v] [--username USERNAME]
 
 Garmin Connect Exporter
 
-options:
+optional arguments:
```
[the remark this answers is not in the document]
Comment (bxsx) [reply README.md:61]: Interesting. Keep it as you like then :)

**Pull request merged**